### PR TITLE
Replaced Job-Structure with source agent cache

### PR DESCRIPTION
**Pull request by paranerd**

### CODE REVIEW

File: server/services/rules-service.ts [hunk 1]
```diff
@@ -129,13 +130,200 @@ export class RulesService {
...
+  }
+
+  /**
+   * Run all rules.
+   */
+  async runAll() {
+    const rules = await rulesService.listRules();
+    const users = collectionService.users;
+
+    for (const rule of rules) {
```
Comment (qbit-42) [inline server/services/rules-service.ts:231]: I think you spoke about it at some point, but why isn't it a filter() / map () chain?

File: server/services/rules-service.ts [hunk 2]
```diff
@@ -129,13 +130,200 @@ export class RulesService {
...
+      // Execute data fetching task.
+      const taskResult = await this.executeSourceAgentTask(
+        rule.source.agentId,
+        sourceTasks
+      );
+
+      logger.info(taskResult);
+
+      // Process task failure.
+      if (taskResult.status !== 'success') {
```
Comment (qbit-42) [inline server/services/rules-service.ts:270]: !== success here, === failed in L290 - care to unify condition logic?

**Review by qbit-42 — APPROVED** [review]
Approved, looks good - with some minor/style comments/questions.

File: server/agents/source-agents/simple-source-agent.ts [hunk 3]
```diff
@@ -72,6 +75,43 @@ export abstract class SimpleSourceAgent implements SourceAgent {
...
+    headers = {}
+  ): Promise<AxiosResponse<any, any>> {
+    // Calculate cache key
+    const cacheKey = createHash('md5')
+      .update(`${url}.${JSON.stringify(params)}`)
+      .digest('hex');
+
+    // Check if request is cached
+    if (Object.keys(this.cache).includes(cacheKey)) {
+      this.logger.info(`Returning cached result!`);
```
[the remark this answers is not in the document]
Comment (echom) [reply server/agents/source-agents/simple-source-agent.ts:101]: And a quick one from my end: Why turn a hash lookup "if(cacheKey in this.cache)" into an array search "Object.keys(...).includes(cacheKey)"?

File: server/services/rules-service.ts (hunk 1 above)
Comment (echom) [reply server/services/rules-service.ts:231]: Filter/map do not handle async/await. Then you'd be map/filtering Promise arrays, which could reduce readability.

**Pull request merged**